### PR TITLE
Add error checking for objects on external heap

### PR DESCRIPTION
**Pull request by philipmarshall21**
Now checking if an object resides on the external heap when determining if said object is symmetric. This PR also revises the overlap check to allow overlap in target and source buffers for RMA operations if `target_pe != source_pe`

### CODE REVIEW

**Comment by philipmarshall21** [comment]
Not convinced the change to SHMEM_ERR_CHECK_SYMMETRIC_HEAP is quite correct yet

**Comment by philipmarshall21** [comment]
Seems like the overlap checks are missing entirely from **data_f.c**, **synchronization_f.c**, and **collectives_f.c**. Are those files still being maintained / does the **SHMEM_ERR_CHECK_OVERLAP** check need to be added to them?

**Comment by davidozog** [comment]
> Seems like the overlap checks are missing entirely from **data_f.c**, **synchronization_f.c**, and **collectives_f.c**. Are those files still being maintained / does the **SHMEM_ERR_CHECK_OVERLAP** check need to be added to them?

That’s probably not a concern, those are for Fortran interfaces, which are all deprecated.

File: src/data_c.c4 [hunk 1]
```diff
@@ -287,7 +287,8 @@ SHMEM_PROF_DEF_CTX_PUT_N_SIGNAL_NBI(`mem')
     SHMEM_ERR_CHECK_SYMMETRIC(target, sizeof(TYPE) * nelems);    \
     SHMEM_ERR_CHECK_NULL(source, nelems);                        \
     SHMEM_ERR_CHECK_OVERLAP(target, source, sizeof(TYPE) *       \
-                            nelems, sizeof(TYPE) * nelems, 0);   \
+                            nelems, sizeof(TYPE) * nelems, 0,    \
+                            (shmem_internal_my_pe == pe));       \
```
Comment (parkerha1) [inline src/data_c.c4:291]: Why do the checks in collectives_c get passed a 1 but these are passing the conditional? Do we know the collectives_c macros are already on the same PE?

File: src/shmem_internal.h [hunk 2]
```diff
@@ -247,46 +247,66 @@ extern hwloc_topology_t shmem_internal_topology;
...
+            }                                                                                 \
+        }                                                                                     \
+        else if (ptr_base >= shmem_internal_heap_base && ptr_base < heap_ext) {               \
+            if (ptr_ext > heap_ext) {                                                         \
+                RAISE_ERROR_MSG("Argument \"%s\" [%p..%p) exceeds "                           \
+                                "sym. heap region [%p..%p)\n", #ptr_in, ptr_base,             \
+                                ptr_ext, shmem_internal_heap_base, heap_ext);                 \
+            }                                                                                 \
+        }                                                                                     \
+        else if (shmem_external_heap_base) {                                                  \
```
Comment (parkerha1) [inline src/shmem_internal.h:275]: what kind of devices could provide an external heap? GPU's would be my first thought, but I was under the impression that SOS doesn't support using their memory.

**Review by parkerha1 — APPROVED** [review]
LGTM

File: src/data_c.c4 (hunk 1 above)
Comment (philipmarshall21) [reply src/data_c.c4:291]: Good question. The "1" value is indicating the source/destination overlap check should always occur for APIs such as the collectives. However, for RMA operations, the source and destination buffers can overlap so long as the destination PE and source PE are different (i.e. `shmem_internal_my_pe == pe` evaluates to 0, ignoring the overlap check).

File: src/shmem_internal.h (hunk 2 above)
Comment (philipmarshall21) [reply src/shmem_internal.h:275]: In the case of Intel® SHMEM, we are using the external heap to allow SOS to interact with data in device (GPU) memory. SOS supports this usage when configured with **--enable-ofi-hmem**.

**Review by davidozog — APPROVED** [review]
Looks great 💯, thanks @philipmarshall21!

**Pull request merged**